crates/ursa-pod/benches/e2e.rs

```diff
@@ -267,5 +292,282 @@ mod http_hyper {
     }
 }
 
+#[cfg(feature = "bench-quic")]
+mod quinn_ufdp {
+    use super::{
+        tls_utils::{client_config, server_config},
+        DummyBackend,
+    };
+    use futures::future::join_all;
+    use quinn::{ConnectionError, Endpoint, RecvStream, SendStream, ServerConfig};
+    use std::{
+        io::Error,
+        pin::Pin,
+        sync::Arc,
+        task::{Context, Poll},
+    };
+    use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+    use tokio::task;
+    use ursa_pod::{client::UfdpClient, server::UfdpHandler, types::Blake3Cid};
+
+    const CLIENT_PUB_KEY: [u8; 48] = [3u8; 48];
+    const CID: Blake3Cid = Blake3Cid([3u8; 32]);
+
+    pub async fn server_loop(
+        addr: String,
+        content: &'static [u8],
+        tx_started: tokio::sync::oneshot::Sender<u16>,
+    ) {
+        let server_config = ServerConfig::with_crypto(Arc::new(server_config()));
+        let server = Endpoint::server(server_config, addr.parse().unwrap()).unwrap();
+        let port = server.local_addr().unwrap().port();
+
+        tx_started.send(port).unwrap();
+
+        while let Some(connecting) = server.accept().await {
+            let connection = connecting.await.unwrap();
+            let content_clone = content.clone();
+            task::spawn(async move {
+                loop {
+                    match connection.accept_bi().await {
+                        Ok((tx, rx)) => {
+                            task::spawn(async {
+                                let stream = BiStream { tx, rx };
+                                let handler = UfdpHandler::new(
+                                    stream,
+                                    DummyBackend {
+                                        content: content_clone,
+                                    },
+                                    0,
+                                );
+                                if let Err(e) = handler.serve().await {
+                                    println!("server error: {e:?}");
+                                }
+                            });
+                        }
+                        Err(ConnectionError::ApplicationClosed(_)) => {
+                            // Client closed the connection.
+                            break;
+                        }
+                        Err(e) => panic!("{e:?}"),
+                    }
+                }
+            });
+        }
+    }
+
+    pub async fn client_loop(addr: String, iterations: usize) {
+        let mut tasks = vec![];
+        let mut endpoint = Endpoint::client("0.0.0.0:0".parse().unwrap()).unwrap();
+        let client_config = quinn::ClientConfig::new(Arc::new(client_config()));
+        endpoint.set_default_client_config(client_config);
+        let stream = endpoint
```

Comment:
@qti3e @ozwaldorf what's nice about using QUIC here is that we don't have to open `iterations` connections like in TCP (if we were using TCP + TLS that would be a lot of connection setup overhead). Instead we can open one connection and send each request in a separate stream. They all get multiplexed in the wire.

Reply:
This is a really great point, and good to consider when we choose the main mode of transport!